### PR TITLE
Task 3 of internship project

### PR DESCRIPTION
**Pull request by AllyMezzen**

### CODE REVIEW

File: test-code-review-main/setWaitingInterval.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+const map = new Map<number, number>();
+
+let waitingIntervalId = 0;
```
Comment (AllyMezzen) [inline test-code-review-main/setWaitingInterval.ts:3]: Better avoid creating global variables (values can be changed by anyone)

File: test-code-review-main/setWaitingInterval.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        const item = arr.pop();
+
+        if(typeof item !== 'number'){
+            throw new Error('Invalid item type');
+        }
+
+        return item;
+    }
+
+    return arr[0];
```
Comment (AllyMezzen) [inline test-code-review-main/setWaitingInterval.ts:16]: Corner-case is not handled: arr can be empty (meaning: function will return undefined)

File: test-code-review-main/setWaitingInterval.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
+const map = new Map<number, number>();
+
+let waitingIntervalId = 0;
+
+function getLastUntilOneLeft(arr: number[]): number {
+    if (arr.length > 1) {
+        const item = arr.pop();
```
Comment (AllyMezzen) [inline test-code-review-main/setWaitingInterval.ts:7]: As arr is passed as a function prop, it shouldn't be mutated (.pop() mutates an original array)

File: test-code-review-main/setWaitingInterval.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * This function mimics the behavior of setInterval with one key difference:
+ * if the callback function takes too long to execute or if the browser throttles,
+ * subsequent calls to the callback function will not occur.
+ *
+ * Additionally, we pass an array of timeouts to define an increasing delay period.
+ * For example, given the array [16, 8, 4, 2], the delays will be 2, 4, 8, 16, 16, 16...
+ */
+export function setWaitingInterval(handler: Function, timeouts: number[], ...args: any[]): number {
```
Comment (AllyMezzen) [inline test-code-review-main/setWaitingInterval.ts:27]: Types Function and any[] don’t specify what is taken or returned (that defeats TypeScript's whole point of type safety)

File: test-code-review-main/setWaitingInterval.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    );
+
+    return waitingIntervalId;
+}
+
+export function clearWaitingInterval(intervalId: number): void {
+    const realTimeoutId = map.get(intervalId);
+
+    if (typeof realTimeoutId === 'number') {
+        clearTimeout(realTimeoutId);
```
Comment (AllyMezzen) [inline test-code-review-main/setWaitingInterval.ts:50]: No cleanup-function for a map is called: each time an interval is cleared it isn't removed from a map causing it to grow further.

**Review by AllyMezzen — COMMENTED** [review]
Current approach is not optimal and has issues that are able to break the functionality.
